cni-plugin/linkerd-cni-validator/src/main.rs

```diff
@@ -0,0 +1,174 @@
+
+    let server_task = tokio::spawn(outbound_serve(outbound_proxy_addr, ready_tx, shutdown_rx));
+
+    let validation_task = tokio::spawn(validate_outbound_redirect(target_addr, ready_rx));
+
+    tokio::select! {
+        task = validation_task => {
+            let validation_result = task.expect("Failed to run validator task");
+            if let Err(err) = validation_result {
+                error!(error = %err, "Failed validation");
+                exit(UNSUCCESSFUL_EXIT_CODE);
+            }
+
+            info!("Validation passed successfully...exiting");
+            Ok(())
+        }
+
+        _ = shutdown_tx.signaled() => {
+            if let Err(e) = server_task.await.expect("Failed to run outbound server task") {
+                error!(error = %e, "Failed to validate outbound routing configuration");
+            } else {
+                error!("Failed to validate due to server terminating early");
+            }
+
+            exit(UNSUCCESSFUL_EXIT_CODE);
+        }
+    }
+}
+
+#[tracing::instrument(
+    level = "debug",
+    skip(ready_rx),
+    fields(original_dst = %target_addr),
+    )]
+async fn validate_outbound_redirect(
+    target_addr: SocketAddr,
+    ready_rx: oneshot::Receiver<Result<()>>,
+) -> Result<()> {
+    let timeout = std::time::Duration::from_secs(120);
+    if tokio::time::timeout(timeout, ready_rx).await.is_err() {
+        anyhow::bail!("timed-out ({:?}) waiting for server to be ready", timeout);
+    }
+
+    let mut stream = {
+        debug!("Building validation client");
+        let socket = TcpStream::connect(target_addr).await?;
+        assert_eq!(target_addr, socket.peer_addr().unwrap());
+        debug!("Client connected to validation server");
+        socket
+    };
+
+    tokio::select! {
+        is_readable = stream.readable() => {
+            is_readable.context("cannot read off client socket")?;
+        }
+
+        () = tokio::time::sleep(timeout) => {
+            anyhow::bail!("timed-out ({:?}) waiting for socket to become readable", timeout);
+        }
+    };
+
+    let mut buf = [0u8; 100];
+    let read_sz = stream.read(&mut buf[..REDIRECT_RESPONSE.len()]).await?;
+    let resp = String::from_utf8(buf[..REDIRECT_RESPONSE.len()].to_vec())?;
+    debug!(redirect_response = %resp, bytes_read = %read_sz);
+    anyhow::ensure!(
+        resp == REDIRECT_RESPONSE,
+        "expected client to receive {:?}, got {:?} instead",
+        REDIRECT_RESPONSE,
+        resp
+    );
+    Ok(())
+}
+
+#[tracing::instrument(name = "outbound_server", skip(ready_tx, shutdown))]
+async fn outbound_serve(
+    listen_addr: SocketAddr,
+    ready_tx: oneshot::Sender<Result<()>>,
+    shutdown: kubert::shutdown::Watch,
+) -> Result<()> {
+    let listener = TcpListener::bind(listen_addr)
+        .await
+        .expect("Failed to bind server");
+    info!("Listening for incoming connections");
+
+    if ready_tx.send(Ok(())).is_err() {
```

Comment:
It looks like we don't ever actually send an error on the channel. Perhaps it should just use a `()`?  Alternatively, we could probably just use an `Arc<Notify>`